File: go/internal/fingerprintcore/fingerprint.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

// Package fingerprintcore contains the stateless error-fingerprint logic
// shared by the top-level telemetry package and the logger sub-package.
package fingerprintcore

import (
	"crypto/sha256"
	"fmt"
	"runtime"
	"strings"
)

// ShortHash12 returns the first 12 hex characters of the SHA-256 hash.
func ShortHash12(input string) string {
	sum := sha256.Sum256([]byte(input))
	return fmt.Sprintf("%x", sum)[:12]
}

// ExtractBasename extracts the filename without path or extension, lowercased.
func ExtractBasename(file string) string {
	file = strings.ReplaceAll(file, "\\", "/")
	if idx := strings.LastIndex(file, "/"); idx >= 0 {
		file = file[idx+1:]
	}
	if idx := strings.LastIndex(file, "."); idx >= 0 {
		file = file[:idx]
	}
	return strings.ToLower(file)
}

// ExtractFuncName extracts just the function name from a fully qualified Go function path.
func ExtractFuncName(fn string) string {
	if idx := strings.LastIndex(fn, "."); idx >= 0 {
		return fn[idx+1:]
	}
	return fn
}

// ComputeFromPCs generates a stable 12-char hex fingerprint from exception type
// + top 3 stack frames from the given program counters.
func ComputeFromPCs(excType string, pcs []uintptr) string {
	parts := []string{strings.ToLower(excType)}
	if len(pcs) > 0 {
		frames := runtime.CallersFrames(pcs)
		count := 0
		for count < 3 {
			frame, more := frames.Next()
			if frame.Function == "" && !more {
				break
			}
			if frame.File != "" {
				base := ExtractBasename(frame.File)
				fn := strings.ToLower(ExtractFuncName(frame.Function))
				parts = append(parts, base+":"+fn)
				count++
			}
			if !more {
				break
			}
		}
	}
	return ShortHash12(strings.Join(parts, ":"))
}

// ComputeFromParts generates a fingerprint from pre-extracted parts.
func ComputeFromParts(excType string, frameParts []string) string {
	parts := []string{strings.ToLower(excType)}
	parts = append(parts, frameParts...)
	return ShortHash12(strings.Join(parts, ":"))
}

File: csharp/src/Provide.Telemetry/Fingerprint.cs
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.


using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace Provide.Telemetry;

/// <summary>
/// Stable error fingerprinting — a 12-char hex hash of the exception type plus
/// the top 3 stack frames, normalized to <c>basename:function</c> with no line
/// numbers so the same failure fingerprints identically across runs.
/// </summary>
/// <remarks>
/// The algorithm is the five-language contract, matching
/// <c>src/provide/telemetry/logger/processors.py</c> (<c>_compute_error_fingerprint</c>),
/// <c>typescript/src/fingerprint.ts</c>, <c>go/internal/fingerprintcore/fingerprint.go</c>, and
/// <c>rust/src/fingerprint.rs</c>: lowercase the parts, join with <c>:</c>,
/// SHA-256 the UTF-8 bytes, and keep the first 12 hex characters.
/// </remarks>
public static class Fingerprint
{
    /// <summary>Number of stack frames contributing to a fingerprint.</summary>
    public const int MaxFrames = 3;

    /// <summary>Hex characters retained from the SHA-256 digest.</summary>
    public const int FingerprintLength = 12;

    // .NET stack trace lines look like:
    //   "   at Namespace.Type.Method() in /path/to/File.cs:line 42"
    //   "   at Namespace.Type.Method()"                    (no PDB / no file info)
    // Group 1 is the fully-qualified method, group 2 the file path when present.
    private static readonly Regex FrameRe = new(
        @"^\s*at\s+(.+?)\s*(?:\(.*?\))?(?:\s+in\s+(.+?):line\s+\d+)?\s*$",
        RegexOptions.Compiled | RegexOptions.CultureInvariant);

    /// <summary>
    /// Compute the fingerprint for an exception type name and optional stack trace.
    /// </summary>
    /// <param name="errorName">Exception type name, e.g. <c>ValueError</c>.</param>
    /// <param name="stackTrace">
    /// Raw <see cref="Exception.StackTrace"/> text, or null when no stack is available.
    /// </param>
    public static string ComputeErrorFingerprint(string errorName, string? stackTrace = null)
        => ComputeErrorFingerprintFromParts(errorName, ExtractFrames(stackTrace));

    /// <summary>
    /// Compute the fingerprint from an exception type name and pre-normalized
    /// <c>basename:function</c> frame strings, mirroring Go's
    /// <c>ComputeErrorFingerprintFromParts</c>.
    /// </summary>
    public static string ComputeErrorFingerprintFromParts(string errorName, IEnumerable<string>? frameParts)
    {
        var parts = new List<string> { errorName.ToLowerInvariant() };
        if (frameParts is not null) parts.AddRange(frameParts);
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes(string.Join(":", parts)));
        return Convert.ToHexStringLower(digest)[..FingerprintLength];
    }

    /// <summary>
    /// Compute the fingerprint directly from an exception.
    /// </summary>
    public static string ComputeErrorFingerprint(Exception exception)
        => ComputeErrorFingerprint(exception.GetType().Name, exception.StackTrace);

    /// <summary>
    /// Normalize the innermost <see cref="MaxFrames"/> stack frames into
    /// <c>basename:function</c> pairs. .NET orders stack traces innermost-first,
    /// so the leading frames are the ones Python and TypeScript keep with
    /// <c>[-3:]</c> / <c>slice(-3)</c> over their outermost-first traces.
    /// </summary>
    internal static List<string> ExtractFrames(string? stackTrace)
    {
        var frames = new List<string>();
        if (string.IsNullOrEmpty(stackTrace)) return frames;

        foreach (var line in stackTrace.Split('\n'))
        {
            var match = FrameRe.Match(line.TrimEnd('\r'));
            if (!match.Success) continue;

            var func = LeafFunction(match.Groups[1].Value);
            var basename = Basename(match.Groups[2].Success ? match.Groups[2].Value : "");
            if (basename.Length == 0) continue;

            frames.Add($"{basename}:{func}");
            if (frames.Count == MaxFrames) break;
        }
        return frames;
    }

    /// <summary>Strip directories and the extension, then lowercase: "/a/b/File.cs" -> "file".</summary>
    private static string Basename(string path)
    {
        var leaf = path.Replace('\\', '/').Split('/')[^1];
        var dot = leaf.LastIndexOf('.');
        return (dot > 0 ? leaf[..dot] : leaf).ToLowerInvariant();
    }

    /// <summary>Keep the final dotted segment of a qualified method name, lowercased.</summary>
    private static string LeafFunction(string qualified)
    {
        var trimmed = qualified.Trim();
        var dot = trimmed.LastIndexOf('.');
        return (dot >= 0 ? trimmed[(dot + 1)..] : trimmed).ToLowerInvariant();
    }
}
